### ISSUE TITLE
[Bug]: CLI cannot handle dev command

### ISSUE THREAD

**Issue by resistorsoftware** [Shopify/cli#1678]
### Please confirm that you have:

- [X] Searched [existing issues](.) to see if your issue is a duplicate. (If you’ve found a duplicate issue, feel free to add additional information in a comment on it.)
- [X] Reproduced the issue in [the latest CLI version](https://www.npmjs.com/package/@shopify/cli).

### In which of these areas are you experiencing a problem?

App

### Expected behavior

Yarn dev should have worked

### Actual behavior

Yarn dev blows chunks on some internal problem, likely due to swapping out one extension for another.

### Verbose output

```shell
Cannot read properties of undefined (reading 'length')                                                                                      │
│                                                                                                                                              │
│  To investigate the issue, examine this stack trace:                                                                                         │
│    at getUIExtensionPayload (@shopify/app/src/cli/services/dev/extension/payload.ts:50)                                                      │
│      metafields: extension.configuration.metafields.length === 0 ? null : extension.configuration.metafields,                                │
│    at all                                                                                                                                    │
│    at async getExtensionsPayloadStoreRawPayload (@shopify/app/src/cli/services/dev/extension/payload/store.ts:40)                            │
│      extensions: await Promise.all(options.extensions.map((extension) => getUIExtensionPayload(extension, options))),                        │
│    at async devUIExtensions (@shopify/app/src/cli/services/dev/extension.ts:96)                                                              │
│      const payloadStoreRawPayload = await getExtensionsPayloadStoreRawPayload(payloadStoreOptions)                                           │
│    at action (@shopify/app/src/cli/services/dev.ts:360)                                                                                      │
│      await devUIExtensions({                                                                                                                 │
│    at action (@shopify/app/src/cli/utilities/app/http-reverse-proxy.ts:63)                                                                   │
│      await target.action(stdout, stderr, signal, targetPort)                                                                                 │
│    at (Users/Documents/Workspace/Apps/2023/foo/node_modules/@shopify/cli-kit/dist/private/node/ui/components/Concurrent  │
│    Output.js:70)                                                                                                                             │
│    at all
```

### Reproduction steps

1.
2.
3.

### Operating System

MacOs

### Shopify CLI version (check your project's `package.json` if you're not sure)

3.44.1

### Shell

zsh

### Node version (run `node -v` if you're not sure)

19.7

### What language and version are you using in your application?

Ruby 3.2.1

**Comment by Arkham** [comment]
Without reproduction steps it's hard to figure out what the problem is. Feel free to reopen if you find some simple replication steps we could follow to debug this.

**Comment by resistorsoftware** [comment]
I love how you close issues where you have a problem, but choose to not even ask for more. I spelled it out, but I will elaborate more. If you swap one extension for another, by replacing code in the extensions directory, you get this error. If on the other hand, you use the CLI to make a new extension, you do not get this error. 

Since there is nothing stopping a person from just making a new extension by hand, and thus getting this error, maybe the CLI could be improved to NOT CHOKE and present this error, but instead, to inform the developer that just presenting extension code is not sufficient, and that one has to use the generate command. 

Just my 2 cents. Clearly, not an issue you care about much, but still, removing friction is important.

**Comment by Arkham** [comment]
Hey David, if you look at the template there is a section that says "Reproduction steps". If you make sure to fill it each time, I won't be closing the issue because of missing reproduction steps next time, ok?

**Comment by Arkham** [comment]
Yup, I think the issue is caused by the fact that we load up your app at the beginning when you run `shopify app dev`, then we assume the directory structure won't change during the execution of the dev server. So in a way you can generate extensions manually, but you need to stop `shopify app dev` and restart the process.

We were discussing on supporting more generic file watchers at some point, so I could see a world where the file watcher tells us "this extension is gone" or "this extension has appeared". If that existed, we could handle this hot-swapping of extensions easily. But alas, for now the only thing I can do is to keep it in mind and surface it the next time :)

**Comment by resistorsoftware** [comment]
The mindset that led to this, was the fact an App has a post-purchase extension. When that was discovered to be useless (only Plus stores can use them), I had to revert to a theme app extension. From my experience with those, I just copied a working theme app extension into my project, after removing the original post-purchase one. BOOM. That was when my **yarn dev** command chooched out and I logged the error for here.
